README.md
# Smart Boiler HASSIO Add-on

This repository implements add-on of smartboiler


[![Open your Home Assistant instance and show the add add-on repository dialog with a specific repository URL pre-filled.](https://my.home-assistant.io/badges/supervisor_add_addon_repository.svg)](https://my.home-assistant.io/redirect/supervisor_add_addon_repository/?repository_url=https%3A%2F%2Fgithub.com%2Fhome-assistant%2Faddons-example)

## Add-ons

This repository contains the following add-ons

### [SmartBoiler add-on](./smartboiler)

![Supports aarch64 Architecture][aarch64-shield]
![Supports amd64 Architecture][amd64-shield]
![Supports armhf Architecture][armhf-shield]
![Supports armv7 Architecture][armv7-shield]
![Supports i386 Architecture][i386-shield]

_SMART BOILER add-on to use as a blueprint for new add-ons._

<!--

Notes to developers after forking or using the github template feature:
- While developing comment out the 'image' key from 'example/config.yaml' to make the supervisor build the addon
  - Remember to put this back when pushing up your changes.
- When you merge to the 'main' branch of your repository a new build will be triggered.
  - Make sure you adjust the 'version' key in 'example/config.yaml' when you do that.
  - Make sure you update 'example/CHANGELOG.md' when you do that.
  - The first time this runs you might need to adjust the image configuration on github container registry to make it public
  - You may also need to adjust the github Actions configuration (Settings > Actions > General > Workflow > Read & Write)
- Adjust the 'image' key in 'example/config.yaml' so it points to your username instead of 'home-assistant'.
  - This is where the build images will be published to.
- Rename the example directory.
  - The 'slug' key in 'example/config.yaml' should match the directory name.
- Adjust all keys/url's that points to 'home-assistant' to now point to your user/fork.
- Share your repository on the forums https://community.home-assistant.io/c/projects/9
- Do awesome stuff!
 -->

[aarch64-shield]: https://img.shields.io/badge/aarch64-yes-green.svg
[amd64-shield]: https://img.shields.io/badge/amd64-yes-green.svg
[armhf-shield]: https://img.shields.io/badge/armhf-yes-green.svg
[armv7-shield]: https://img.shields.io/badge/armv7-yes-green.svg
[i386-shield]: https://img.shields.io/badge/i386-yes-green.svg


smartboiler/CHANGELOG.md
<!-- https://developers.home-assistant.io/docs/add-ons/presentation#keeping-a-changelog -->

## 0.0.1.3

- Initial release


## 0.0.1.6
- updated files

src/smart_boiler/boiler.py
##########################################################
# Bachelor's thesis                                      #
# From a dumb boiler to a smart one using a smart socket #
# Author: Adam Grünwald                                  #
# BUT FIT BRNO, Faculty of Information Technology        #
# 26/6/2021                                              #
#                                                        #
# Module with class representing boiler itself.          #
# Serves for computing time needed to heat to certain    #
# temperature or recalculates number of showers          #
# to needed temperature to heat to.                      #
##########################################################

from datetime import datetime, timedelta
from requests import get
import pandas as pd


class Boiler:
    def __init__(self, bearer_token, capacity=100, wattage=2000, set_tmp=60, one_shower_volume=40, shower_temperature=40, min_tmp=37, hass_url = "http://localhost:8123"):

        print("------------------------------------------------------\n")
        print('initializing of control...\n\tCapacity of Boiler = {}\n\t Wattage of boiler = {}\n'.format(
            capacity, wattage))
        print("------------------------------------------------------\n")

        self.boiler_heat_cap = capacity * 1.163
        self.real_wattage = wattage * 0.98
        self.set_tmp = set_tmp
        self.capacity = capacity
        self.one_shower_volume = one_shower_volume
        self.shower_temperature = shower_temperature
        self.min_tmp = min_tmp
        self.power_on = False
        self.entity_id = None
        self.hass_url = hass_url
        self.bearer_token = bearer_token
        
    def control_power(self):

        url = "http://localhost:8123/ENDPOINT"
        headers = {
            "Authorization": self.bearer_token,
            "content-type": "application/json",
        }

        response = get(self.hass_url, headers=headers)
        print(response.text)

    def time_needed_to_heat_up(self, tmp_change):
        """
        Calculates the time needed for heating water in boiler by temperature delta.
            time = (m * c) * d'(tmp) / P * effectivity_coef

            https://vytapeni.tzb-info.cz/tabulky-a-vypocty/97-vypocet-doby-ohrevu-teple-vody
        """

        return (self.boiler_heat_cap * tmp_change) / (self.real_wattage)

    def is_needed_to_heat(self, tmp_act, tmp_goal, time_to_consumption):
        """Conciders if it is needed to heat.

        Args:
            tmp_act ([type]): [actual temperature of water in boiler]
            tmp_goal ([type]): [temperature of water to heat before consumption]
            time_to_consumption ([type]): [time to consumption]

        Returns:
            [boolean]: [boolean value of needed to heat]
        """
        tmp_change = tmp_goal - tmp_act

        if (tmp_change > 0) and (time_to_consumption <= self.time_needed_to_heat_up(tmp_change)):
            return True
        else:
            return False

    def showers_degrees(self, number_of_showers):
        """Recalculates number of showers to temperature 
        on which is needed to heat up water in boiler.

        Args:
            number_of_showers ([int]): [number of showers to prepare]

        Returns:
            [float]: [tempreture to heat on]
        """
        showers_volume = number_of_showers * self.one_shower_volume

        cold_water_tmp = 10

        needed_temperature = ((self.min_tmp * self.capacity) + (self.shower_temperature *
                              showers_volume) - (showers_volume * cold_water_tmp)) / self.capacity

        if needed_temperature > self.set_tmp:
            return self.set_tmp

        return needed_temperature

    def real_tmp(self, tmp_act):
        """Calculates the real temperature of water in boiler

        Args:
            tmp_act ([float]): [actual temperature measured in boiler]

        Returns:
            [float]: [the real temperature of water in boiler]
        """

        if((tmp_act is None) or (self.area_tmp is None) or (self.set_tmp is None)):
            return 50

        if(tmp_act < self.area_tmp or tmp_act > self.set_tmp):
            return tmp_act

        tmp_act_and_area_delta = tmp_act - self.area_tmp
        tmp_max_and_area_delta = self.boiler_measured_max - self.area_tmp

        p1 = tmp_act_and_area_delta / tmp_max_and_area_delta

        tmp = p1 * (self.set_tmp - self.area_tmp) + self.area_tmp
        return tmp

    def set_measured_tmp(self, df):
        """Initializes measured temperatures for calculation of real temperature of water in boiler.

        Args:
            df ([DataFrame]): [dataframe with measured data]
        """
        df_of_last_week = df[df.index > (
            df.last_valid_index() - timedelta(days=21))]

        # self.area_tmp = df_of_last_week['tmp1'].nsmallest(100).mean()
        self.area_tmp = 20
        self.boiler_measured_max = df_of_last_week['tmp2'].nlargest(100).mean()

        print("area_tmp: ", self.area_tmp,
              "\nboiler_max: ", self.boiler_measured_max)


if __name__ == '__main__':
    tmp_act = 30
    tmp_goal = 52
    time_to_consumption = 1
    b = Boiler(capacity=80, wattage=2000)
    print(b.is_needed_to_heat(tmp_act, tmp_goal, time_to_consumption))


src/smart_boiler/control.py
###########################################################
# Bachelor's thesis                                       #
# From a dumb boiler to a smart one using a smart socket  #
# Author: Adam Grünwald                                   #
# BUT FIT BRNO, Faculty of Information Technology         #
# 26/6/2021                                               #
#                                                         #
# Module that controls a heating of smart boiler.         #
# Uses module Boiler for computing time needed to heating #
# of water, module TimeHandler for basic time and date    #
# operations, WeekPlanner for plan week heating,          # 
# SettingsLoader to load setting from settings file and   # 
# EventChecker which checks events in calendar,           # 
# when the water shouldn't be heated.                      #
###########################################################


from datetime import datetime, timedelta
import pandas as pd
import calendar
from influxdb import DataFrameClient
from influxdb import InfluxDBClient
import os.path
import signal
import sys

import time
import json
import requests
import homeassistant.remote as remote



from scipy.misc import electrocardiogram
import numpy as np


from boiler import Boiler
from time_handler import TimeHandler
from week_planner import WeekPlanner
from event_checker import EventChecker


class Controller:
    """Main class which makes decisions about about heating
    """

    def __init__(self, settings_file='settings.json'):
        """Inits class of Controller. Loads settings from a settings file

        Args:
            settings_file (str, optional): [name of json file with settings]. Defaults to 'settings.json'.
        """

        from settings_loader import SettingsLoader
        SettingsLoader = SettingsLoader(settings_file)
        settings = SettingsLoader.load_settings()

        self.host = settings['db_host']
        self.port = settings['db_port']
        self.db_name = settings['db_name']
        self.measurement = settings['measurement']
        self.db_user = settings['db_user']
        self.db_pass = settings['db_pass']
        self.socket_url = settings['socket_url']

        self.tmp_output = settings['tmp_output']
        self.tmp_boiler = settings['tmp_boiler']

        self.tmp_min = settings['tmp_min']
        self.consumption_tmp_min = settings['consumption_tmp_min']

        self.start_date = datetime.now()
        
        self.Hass = remote.API('localhost', 'smart_boiler01')
        self.shelly_entity_id = 'shelly1pm_34945475a969'


        boiler_wattage = settings['boiler_wattage']
        boiler_capacity = settings['boiler_capacity']
        boiler_set_tmp = settings['boiler_set_tmp']

        print("------------------------------------------------------\n")
        print('initializing of Control...\n\tdb_name = {}\n\tsocker_url = {}\n\ttmp_output = {}\n\ttmp_boiler = {}\n\thost name = {}\n\tport = {}\n\tboiler capacity = {}\n\tboiler woltage = {}\n'.format(
            self.db_name, self.socket_url, self.tmp_output, self.tmp_boiler, self.host, self.port, boiler_capacity, boiler_wattage))
        print("------------------------------------------------------\n")

        self.InfluxDBClient = InfluxDBClient(
            host=self.host, 
            port=self.port, 
            username=self.db_user, 
            password=self.db_pass, 
            retries=5, 
            timeout=1)
        self.DataFrameClient = DataFrameClient(
            username=self.db_user, 
            password=self.db_pass,
            host=self.host, 
            database=self.db_name)

        #self.EventChecker = EventChecker()
        #self.TimeHandler = TimeHandler()
        #self.Boiler = Boiler(capacity=boiler_capacity,
        #                    wattage=boiler_wattage, set_tmp=boiler_set_tmp)

        #self.data_db = self._actualize_data()
        self.last_data_update = datetime.now()
        self.last_legionella_heating = datetime.now()

        #self.WeekPlanner = WeekPlanner(self.data_db)
        self.coef_up_in_current_heating_cycle_changed = False
        self.coef_down_in_current_heating_cycle_changed = False

    def _last_entry(self):
        """Loads last entru from DB - actual.

        Args:
            measurement (str, optional): [description]. Defaults to 'senzory_boiler'.

        Returns:
            [type]: [last entry values]
        """
        try:
            result = self.InfluxDBClient.query(
                'SELECT * FROM "' + self.db_name + '"."autogen"."' + self.measurement + '" ORDER BY DESC LIMIT 1')

            result_list = list(result.get_points(
                measurement=self.measurement))[0]

            time_of_last_entry = self.TimeHandler.date_from_influxdb_to_datetime(
                result_list["time"])
            tmp1 = result_list["tmp1"]
            tmp2 = result_list["tmp2"]
            socket_turned_on = result_list["turned"]

            return {"tmp1": tmp1, "tmp2": tmp2, "socket_turned_on": socket_turned_on, "time_of_last_entry": time_of_last_entry}

        except:
            print("unable to read from influxDBclient")
            return None

    def _actualize_data(self):
        """
        Actualizes data form DB

        Args:
            measurement (str, optional): [description]. Defaults to 'senzory_boiler'.
            host (str, optional): [description]. Defaults to 'influx_db'.

        Returns:
            [type]: Returns data from DB
        """
        print("trying to get new datasets...")

        try:

            datasets = self.DataFrameClient.query(
                # 'SELECT * FROM "' + self.db_name + '"."autogen"."' + self.measurement + '" ORDER BY DESC')[self.measurement]
                'SELECT * FROM "' + self.db_name + '"."autogen"."°C" ORDER BY DESC')['°C']

            print("got new datasets")
            df = pd.DataFrame(datasets)
            # df = df[df.in_event != True]
            self.Boiler.set_measured_tmp(df)

            return df

        except:
            print("it wasnt possible to get new datasets")
            return None


    def _check_data(self):
        """ Refreshs data every day
        """
        if self.last_data_update - datetime.now() > timedelta(days=1):
            print(datetime.now())
            print("actualizing data")

            actualized_data = self._actualize_data()
            if actualized_data is not None:
                self.data_db = actualized_data
                self.WeekPlanner.week_plan(self.data_db)

    def _learning(self):
        """ After one week of only measuring the data starts heating based on historical data.

        Returns:
            [boolean]: [True if in learing]
        """

        return ( ( datetime.now() - self.start_date) < timedelta(days=7) )

    def control(self):
        """ Method which decides about turning on or off the heating of a boiler.
        """

        self._check_data()

        last_entry = self._last_entry()

        # checks whether the water in boiler should be even ready 
        if self.EventChecker.check_off_event():
            print("naplanovana udalost")
            self.toggle_shelly_relay('off')
            time.sleep(600)
            return

        # last measured entry in DB
        if last_entry is None:
            self.toggle_shelly_relay('on')
            return


        # looks for the next heat up event from a calendar    
        next_calendar_heat_up_event = self.EventChecker.next_calendar_heat_up_event(
            self.Boiler)

        time_now = datetime.now()
        day_of_week = time_now.weekday()

        # actual tmp of water in boiler
        tmp_act = self.Boiler.real_tmp(last_entry['tmp2'])

        # state of smart socket
        is_on = last_entry['socket_turned_on']
        # in first week is water in boiler hold around 60 degrees

        #protection from freezing
        if tmp_act < 5:
            if not is_on:
                snf.

        if self._learning():
            if tmp_act > 60:
                if is_on:
                    self.toggle_shelly_relay('off')
            else:
                if tmp_act < 57:
                    if not is_on:
                        self.toggle_shelly_relay('on')

            return


        # if last entry is older than 10 minutes and not because of high tarif, water in a boiler is heated for sure
        time_of_last_entry = last_entry['time_of_last_entry']
        if (time_now - time_of_last_entry > timedelta(minutes=10)):
            if not self.WeekPlanner.is_in_DTO():
                print("too old last entry ({}), need to heat".format(
                    time_of_last_entry))
                if not is_on:
                    snf.
            return


        # helping variables for changing day coefs
        if(self.WeekPlanner.is_in_heating()):
            self.coef_down_in_current_heating_cycle_changed = False
        else:
            self.coef_up_in_current_heating_cycle_changed = False


        # once a three weeks is water in boiler heated on max for ellimination of Legionella
        if self.last_legionella_heating - datetime.now() > timedelta(days=21):
            self.coef_down_in_current_heating_cycle_changed = True
            if not is_on:
                print(
                    "starting heating for reduce legionella, this occurs every 3 weeks")
                snf.

            if tmp_act >= (65):
                time.sleep(1200)
                self.last_legionella_heating = datetime.now()
                print("legionella was eliminated, see you in 3 weeks")

        # if is scheduled event for heating, there is evaluated if it is needed to heat
        if next_calendar_heat_up_event['hours_to_event'] is not None:
            time_to_without_DTO = self.WeekPlanner.duration_of_low_tarif_to_next_heating(
                next_calendar_heat_up_event['hours_to_event'])
            tmp_goal = next_calendar_heat_up_event['degree_target']
            print("time to next heating without DTO: ", time_to_without_DTO)
            print("tmp goal : ", tmp_goal)
            if(self.Boiler.is_needed_to_heat(tmp_act, tmp_goal=tmp_goal, time_to_consumption=time_to_without_DTO)):
                print("planned event to heat up with target {} Celsius occurs in {} hours".format(
                    next_calendar_heat_up_event['degree_target'], next_calendar_heat_up_event['hours_to_event']))
                if not is_on:
                    snf.
                return

        # if is actual tmp lower than tmp min, there is need to heat
        if (tmp_act < self.tmp_min):
            if not is_on:
                snf.
            return


        if (self.WeekPlanner.is_in_heating()):

            current_heating = self.WeekPlanner.next_heating_event('end')
            if (current_heating is None):
                return
            current_heating_half_duration = current_heating['duration'] / 2
            how_long_to_current_heating_end = current_heating['will_occur_in']
            
            if(tmp_act < self.consumption_tmp_min):

                print("in heating, needed to increase tmp({}°C) above tmp min({}°C)".format(
                    tmp_act, self.consumption_tmp_min))
                if not is_on:
                    snf.

                if((how_long_to_current_heating_end > current_heating_half_duration) and not self.coef_up_in_current_heating_cycle_changed):
                    #changing the coefs if the temperature during predicted consumption is too low
                    self.coef_up_in_current_heating_cycle_changed = True
                    self.WeekPlanner.week_days_coefs[day_of_week] *= 1.015
                    print("changing day ({}) coef to {}".format(
                        (day_of_week + 1), self.WeekPlanner.week_days_coefs[day_of_week]))

            else:
                if is_on:
                    print("turning off in heating, actual_tmp = {}".format(tmp_act))

                    self.to

            return
        else:
            # checking whether it is needed to heat before the next predicted consumption
            next_heating = self.WeekPlanner. next_heating_event('start')
            if (next_heating is None):
                return
            time_to_next_heating = self.WeekPlanner.duration_of_low_tarif_to_next_heating(
                next_heating['will_occur_in'])

            next_heating_goal_temperature = next_heating['peak'] * \
                self.WeekPlanner.week_days_coefs[day_of_week]

            if(self.Boiler.is_needed_to_heat(tmp_act, tmp_goal=next_heating_goal_temperature, time_to_consumption=time_to_next_heating)):
                print("need to heat up before consumption, time to coms:{} , time without DTO: {}".format(
                    next_heating['will_occur_in'], time_to_next_heating))
                if not is_on:
                    print("boiler is needed to heat up from {} to {}. turning socket on".format(
                        tmp_act, next_heating_goal_temperature))
                    snf.

                return
            # te day coef is changed whether the temperature is too high outside of the predicted consumption
            if ((tmp_act > (self.consumption_tmp_min + 3)) and not self.coef_down_in_current_heating_cycle_changed):
                self.coef_down_in_current_heating_cycle_changed = True
                print("actual tmp is greater than consumption tmp min")
                self.WeekPlanner.week_days_coefs[day_of_week] *= 0.985
                print("changing day ({}) coef to {}".format(
                    (day_of_week + 1), self.WeekPlanner.week_days_coefs[day_of_week]))

            # if boiler need to heat tmp act, tmp act + delta, time to next high tarif
            next_high_tarif_interval = self.WeekPlanner.next_high_tarif_interval(
                'start')
            if next_high_tarif_interval is not None:
                tmp_delta = next_high_tarif_interval['tmp_delta']
                if tmp_delta > 0:
                    time_to_next_high_tarif_interval = next_high_tarif_interval['next_high_tarif_in']
                    if (self.Boiler.is_needed_to_heat(tmp_act, tmp_goal=tmp_act + tmp_delta, time_to_consumption=time_to_next_high_tarif_interval)):
                        if not is_on:
                            print("heating up before in high tarif consumption from {} to {}".format(
                                tmp_act, tmp_act + tmp_delta))
                        return
            if is_on:
                print("turning off outside of heating, actual_tmp = {}".format(tmp_act))
                self.to

    
    def toggle_shelly_relay(self, action):
        service = 'switch.turn_' + action
        data = {'entity_id': self.shelly_entity_id}
        self.Hass.services.call('switch', service, data)


if __name__ == '__main__':

    import sys

    from optparse import OptionParser
    parser = OptionParser('%prog [OPTIONS] <host> <port>')

    parser.add_option(
        '-f', '--settings_file', dest='settings_file',
        type='string',
        default=None
    )
    options, args = parser.parse_args()

    settings_file = options.settings_file

    c = Controller(settings_file)

    while (1):
        # c.control()
        c.toggle_shelly_relay('on')
        time.sleep(60)
        c.toggle_shelly_relay('off')
        
        time.sleep(60)

